README.md
# poly
[![Pub](https://img.shields.io/pub/v/poly.svg)](https://pub.dartlang.org/packages/poly)
[![build status](https://travis-ci.org/Sacchid/poly.svg)](https://travis-ci.org/Sacchid/poly#)
[![License](https://img.shields.io/badge/License-BSD%202--Clause-blue.svg)](https://opensource.org/licenses/BSD-2-Clause)

A library for checking if given point(s) is present inside Polygon or not.

### Contents
*   [Installation](#installation)
*   [Examples](#examples)
*   [Note: Instead of casting, use `toListNum()` & `toListListNum()`](#note-instead-of-casting-use-tolistnum--tolistlistnum)
*   [Function List](#function-list)
    - [Conversions ](#conversion-type)  
    - [is Point(s) inside](#is-points-inside)
        1. [Check if Single `Point` is inside](#check-if-single-point-is-inside)      
        2. [Check if Multiple Points are inside given Polygon](#check-if-multiple-points-are-inside-given-polygon)        
    - [Checks if 2 `Polygon` have same vertices i.e. `points`](#checks-if-2-polygon-have-same-vertices-ie-points)   
    - [CSV](#csv)   
* [Exceptions](#exceptions)
* [Detailed Index](#index)

# Installation

[To install poly for your Dart project, simply add poly to your
pub dependencies.](https://pub.dartlang.org/packages/poly#-installing-tab- "installing poly from pub.dartlang")

    dependencies:
        poly^ 1.0.5

## Examples
### [1. A  very simple usage example ](example/poly_example.dart "Basic example")
 1. Creates 2 Polygons from `List<Point>`
 2. Checks if 2 `Polygon` have same vertices i.e. `points` : using `hasSamePoint()`
 2. Prints if given Points are inside Polygon
 3. Prints if all points in list are inside Polygon or not 
 4. Prints list of result for List of points if they are inside Polygon or not
* Here `hasSamePoint()`, `isPointInside(Point( ))`, `contains(x,y)`, `areAllPointsInsidePolygon_ListPoint()` & `getList_IsListOfPointInside()` are used.

### [2. Example of Conversions `List <=> Point`, `List<Point> <=> List<List>` & other](example/conversion.dart) 
1. Example of `toPoint()` & `pointToList()`
   * `toPoint()` converts `List<num>` to `Point<num>`
    * `pointToList()` converts `Point<num>` to `List<num>`
2. Example of `toListOfPoint()` & `pointsToList()`
   * `toListOfPoint()` converts `List<List<num>>` to `List<Point<num>>`
   * `pointsToList()` converts `List<Point<num>>` to `List<List<num>>`
3. Example of `toPolyFromListOfList()` & `listOfList()`
   * `toPolyFromListOfList()` converts `List<List<num>>` to `Polygon`
   * `listOfList()` returns `Polygon.points` as `List<List<num>>`
   * `toPolyFromListOfList()` gives Polygon & `backToListC2()` gives back List<List<num>>
   * Print status of List of List if they are inside our Polygon using `getList_IsListOfListInside`
   * Print if All points in List of List inside Polygon using


### [3. Examples of `List<dynamic>` => `List<num>` and `List<List<dynamic>>` => `List<List<num>>`](example/using_to.dart)
* `toListNum()` returns `List<num>` from a `List<dynamic>`
*  `toListListNum()` returns `List<List<num>>` from a `List<List<dynamic>>`
1. Examples of `toListNum()` 
    - without any optional parameters
    - with `replaceWithZero: true` and `sizeTwo: false`
    - with `sizeTwo: false`
2. Examples of `toListListNum()` 
    - without any optional parameters
    - with `swapXAndY: true`
    - with `replaceWithZero: true`
    - with `replaceWithZero: true` and `swapXAndY: true`

### [4. Simple CSV example ](example/simple_csv.dart)
* It contains examples of following functions - 
1. Example of `IsInsideResultWithXY_ToCSVString` 
   * Saving `ArePointsInside Results` to "IsInside.csv"
     * Headers row will be added, as `includeHeader` isn`t passed as `false`
     * And as `diffNameThanIsInside:"Example String"` is passed,
     * Header row will be `latitude,longitude,Example String`

2. Example of `toCSVString`
   * Saving `Polygon.points` to "Polygon.csv"
     * Headers row will be added, as `includeHeader` isn`t passed as `false`
     * And as `useXY` is passed as `true`
     * Header row will be `x,y`

3. Example of `csvToResult`
    * Display 1st row : 2nd element & 3rd element of "IsInside.csv"
        * e.g. here "longitude" and "Example String"
        * As, previously `xY_IsInside_ToCSVString` returned String with header
        * because optional parameter `header` was not set to false
   
4. Example of `csvToPoly`
    * Check if Point(18.507305, 73.806131) is inside Polygon readPolygon (Polygon from Polygon.csv)


### [5. Exception Handling Example](example/exception_handling.dart)
* It contains examples of following exceptions & errors -
1. `NeedsAtLeastThreePoints` is thrown if `Polygon.points` contains less than 3 points
2. `WrongSizeForPoint` is thrown if `toPoint()` has more or less than 2 element. Point has only x and y.
3. **`TypeError` is thrown if `List<dynamic>` is passed instead of `List<num>`** 
    * Here, casting can be used to cast `List<dynamic>` to `List<num>`
    * e.g. ` print(toPoint([1,2]));` 
      * Here, [1,2] has a type `List<dynamic>`
      * So, use `[1,2].cast<num>()`
4. `CastError example` - wrongly casting `List<num>` to `List<List<num>`


### [6. Easy Casting Example](example/easy_casting.dart)
* Without casting `List<dynamic>` to `List<num>` `TypeError`is thrown
1. Correct casting
    * casting `List<dynamic>` to `List<num>`
    * casting `List<dynamic>` to `List<List<num>>`
 
2. Passing `List` instead of `List<List>`
* Passing `List` instead of `List<List>` & casting it : throws `CastError` as shown below :
    > type `int` is not a subtype of type `List<num>` in type cast
* Passing `List` instead of `List<List>` but, without casting it : throws `TypeError` as shown below :
    > type `List<dynamic>` is not a subtype of type `List<List<num>>`
* [****Note: currently casting `List<List<dynamic>>` to `List<List<num>>` gives following `CastError` exception:****](https://github.com/dart-lang/sdk/issues/36614 "Dart-lang List<List<dyanamic>> to List<List<num>> Casting Issue")
    * without `as List<List<num>>` - 
    > type `List<dynamic>` is not a subtype of type `List<num>` in type cast
    * with `as List<List<num>>` - 
    > type `List<List<dynamic>>` is not a subtype of type `List<List<num>>` in type cast


## Note: Instead of casting, use `toListNum()` & `toListListNum()`
* Use `toListNum()` for `List<dynamic>` => `List<num>`
* Use `toListListNum()` for`List<List<dynamic>>` => `List<List<num>>`
        

# Function List

### Conversion Type
##### `List<num>` to `Point(x,y)` : use `toPoint()`
* i.e. `[x,y]`  ->  `Point(x,y)`
* Point can be created passing `List<num>` `toPoint()`.
* List must have exact 2 elements, else will throw `WrongSizeForPoint` exception

##### `Point(x,y)` to `List<num>` : use `pointToList()`
* i.e.  `Point(x,y)` ->  `[x,y]`
* List can be created passing `Point(x,y)` `pointToList()`.

##### `List<List<num>>` to `List<Point<num>>` : use `toListOfPoint()`
* i.e. `[ [x1,y1],[x2,y2],... ]`  ->  `[ Point(x1,y1), Point(x2,y2),... ]`
* List of Points can be created from `List<List<num>>` by passing it to `toListOfPoint()`

##### `List<Point<num>>` to `List<List<num>>` : use `pointsToList()`
* i.e.  `[ Point(x1,y1), Point(x2,y2),... ]` ->  `[ [x1,y1],[x2,y2],... ]`
* List can be created passing `List<Point<num>>` `pointsToList()`.

##### `List<List<num>>` to `Polygon` : use `toPolyFromListOfList()`
* i.e. `[ [x1,y1],[x2,y2],... ]`  ->  `Polygon( Point(x1,y1), Point(x2,y2),... )`
* Polygon can be returned from `List<List<num>>` by passing it to `toPolyFromListOfList()`

##### `List<List<num>>` to `Polygon` : use `listOfList()`
* i.e. `[ [x1,y1],[x2,y2],... ]`  from  `Polygon( Point(x1,y1), Point(x2,y2),... )` 
* `List<List<num>>` can be returned from `Polygon` by passing it to `listOfList()`

 
##### `List<dynamic>` to `List<num>` : use `toListNum()`
 * Returns `List<num>` from a `List<dynamic>`
 * Can be used with [toPoly] as it accepts `List<num>`
 * Optional Parameters -
   * `sizeTwo`
     - Default value `true`
     - When set `false`, `Output List` can have more than 2 elements
   * `replaceWithZero`
     - Default value `false`
     - When set `true`, elements with type `String` or `bool` will be replaced with 0, rather than being removed
   * `reverseIt`
     - Default value `false`
     - When set `true`, `List` will be reversed
     
##### `List<List<dynamic>>` to `List<List<num>>` : use `toListListNum()`
 * Returns `List<List<num>>` from a `List<List<dynamic>>`
 * Can be used with functions like [areAllPointsInsidePolygon_List] , [getList_IsListOfListInside] , [toPolyFromListOfList] , [toListOfPoint]  which accepts `List<List<num>>`
 * Optional Parameters -
   * `replaceWithZero`
     - Default value `false`
     - When set `true`, elements with type `String` or `bool` will be replaced with 0, rather than being removed
   * `swapXAndY`
     - Default value `false`
     - When set `true`, `xi` will be swapped with `yi` 
       - i.e. `[ [x1,y1], [x2,y2], ...]` -> `[ [y1,x1], [y2,x2], ...]`

### is Point(s) inside
##### Check if Single `Point` is inside
  ###### Get Status by passing `x` and `y` to `contains`
  * returns `true` if `(x,y)` is present inside `Polygon`
  ###### Get Status by passing `Point(x,y)` to `isPointInside`
  * returns `true` if `Point` is present inside `Polygon`

##### Check if Single `Point` is inside Polygon with tolerance of T meter
  ###### Get Status by passing `Point(x,y)` and Tolerance `T` to `isPointInside`
  
##### Check if Multiple Points are inside given Polygon
 ###### Get Status of each Point
* `getList_IsListOfListInside(List<List<num>>)` returns `List<bool>` 
* `getList_IsListOfPointInside(List<Point<num>>)` returns `List<bool>` 
 ###### Check if all given Points are inside given Polygon
* `areAllPointsInsidePolygon_List((List<List<num>>)` returns `true` or `false`
* `areAllPointsInsidePolygon_ListPoint(List<Point<num>>)` returns `true` or `false`

### Checks if 2 `Polygon` have same vertices i.e. `points`
* use `hasSamePoints()`


### CSV
#### `Get result(s) along with lat, lang as a CSV String : IsInsideResultWithXY_ToCSVString()`
* Returns result of `ArePointsInside` as CSV String which can be later saved or displayed
* Output CSV String will by default contain a header row - `latitude,longitude,isInside`
* Optional parameter: `bool useXY` 
     * By passing, optional parameter: `useXY` as true, header will be `x,y` instead of `latitude,longitude`
     * Default value of `useXY` is `false`
* Optional parameter: `String includeHeader`
     * if optional parameter - `includeHeader` is passed as `false`, returning String will not contain header row
 * Optional Named parameter: `String diffNameThanIsInside`
     * Different name than Default name(`isInside`) will be used by passing optional parameter: `diffNameThanIsInside`
#### `Get `Polygon` as CSV String : toCSVString()`
 * Returns `Polygon` as CSV String which can be later saved or displayed
 * Output CSV String will by default contain a header row - `latitude,longitude`
 * Optional Named parameter: `bool useXY`
    * By passing, optional parameter: `useXY` as true, header will be `x,y` instead of `latitude,longitude`
    * Default value of `useXY` is `false`
 * Optional Named parameter: `String includeHeader`
    * if optional parameter - `includeHeader` is is passed as `false`, returning String will not contain header row
#### `Get `Future<Polygon>` based on `csvString` : csvToPoly()`
 * Returns `Future<Polygon>` based on `csvString`
 * `csvString` may or may not contain header row
 * This function checks if `latitude,longitude` or `x,y` are reversed
     * By checking Header row label
     * i.e. By checking 1st row 1st element is neither "longitude" or "y"
   * If they are reversed, Returned `Polygon` will be `Polygon(latitude,longitude)`, instead of `Polygon(longitude,latitude)`
   * This can be manually set by passing optional parameter: `isReversed` as `true`
 * Optional parameter: `isReversed`
   * `isReversed` has default value = `false`
#### `csvToListOfList()`
 * Returns `Future<List<List>>` based on `csvString`
    * which then can be used - convert that list into `Polygon`
 * Optional parameter: `bool noHeader`
     * By passing optional parameter: `noHeader` as `true`, Resulting List will not contain header row
     * Default value `false`


### Exceptions
1. `NeedsAtLeastThreePoints` is thrown if `Polygon.points` contains less than 3 points
2. `WrongSizeForPoint` is thrown if `toPoint()` has more or less than 2 element. Point has only x and y.
3. **`_TypeError` is thrown if `List<dynamic>` is passed instead of `List<num>`** 
    * Here, casting can be used to cast `List<dynamic>` to `List<num>`
    * e.g. ` print(toPoint([1,2]));` 
      * Here, [1,2] has a type `List<dynamic>`
      * So, use `[1,2].cast<num>()`
4. `_CastError example` - casting `List<num>` to `List<List<num>`

### Index
*   [Installation](#installation)
*   [Examples](#examples)
    - [1. A  very simple usage example ](#1-a--very-simple-usage-example-examplepoly_exampledart-basic-example)
    - [2. Example of Conversions : `List <=> Point`, `List<Point> <=> List<List>` & other](#2-example-of-conversions-list--point-listpoint--listlist--otherexampleconversiondart)
    - [3. Examples of `List<dynamic>` => `List<num>` and `List<List<dynamic>>` => `List<List<num>>`](#3-examples-of-listdynamic--listnum-and-listlistdynamic--listlistnumexampleusing_todart)
    - [4. Simple CSV example](#4-simple-csv-example-examplesimple_csvdart)
    - [5. Exception Handling Example](#5-exception-handling-exampleexampleexception_handlingdart)
    - [6. Easy Casting Example](#6-easy-casting-exampleexampleeasy_castingdart)
*   [Note: Instead of casting, use `toListNum()` & `toListListNum()`](#note-instead-of-casting-use-tolistnum--tolistlistnum)
*   [Function List](#function-list)
    - [Conversions ](#conversion-type)
        1. [`toPoint()`](#listnum-to-pointxy--use-topoint)
        2. [`toListOfPoint()`](#listlistnum-to-listpointnum--use-tolistofpoint)
        3. [`toPolyFromListOfList()`](#listlistnum-to-polygon--use-topolyfromlistoflist)
        4. [`toListNum()`](#listdynamic-to-listnum--use-tolistnum)
        5. [`toListListNum()`](#listlistdynamic-to-listlistnum--use-tolistlistnum)
    
    - [is Point(s) inside](#is-points-inside)
        1. [Check if Single `Point` is inside](#check-if-single-point-is-inside)
            * [Get Status by passing `x` and `y` to `contains`](#get-status-by-passing-x-and-y-to-contains)
            * [Get Status by passing `Point(x,y)` to `isPointInside`](#get-status-by-passing-pointxy-to-ispointinside)
        
        2. [Check if Multiple Points are inside given Polygon](#check-if-multiple-points-are-inside-given-polygon)
            * [Get Status of each Point](#get-status-of-each-point)
            * [Check if all given Points are inside given Polygon](#check-if-all-given-points-are-inside-given-polygon)
   
    - [Checks if 2 `Polygon` have same vertices i.e. `points`](#checks-if-2-polygon-have-same-vertices-ie-points)
    
    - [CSV](#csv)
        * [`Get result(s) along with lat, lang as a CSV String : IsInsideResultWithXY_ToCSVString()`](#get-results-along-with-lat-lang-as-a-csv-string--isinsideresultwithxy_tocsvstring)
        * [`Get `Polygon` as CSV String : toCSVString()`](#get-polygon-as-csv-string--tocsvstring)
        * [`Get `Future<Polygon>` based on `csvString` : csvToPoly()`](#get-futurepolygon-based-on-csvstring--csvtopoly)
        * [`csvToListOfList()`](#csvtolistoflist)

* [Exceptions](#exceptions)

* [Index](#index)

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/Sacchid/poly/issues

## Licence 
Implemented `contains` function logic from [StageXL - A fast and universal 2D rendering engine for HTML5 and Dart](https://github.com/bp74/StageXL) 

As, StageXL imports ``dart:html``, it can not be used in [console application](https://www.dartlang.org/tutorials/server/cmdline) or in [aqueduct back-end](https://aqueduct.io/).

Also, function `distanceInMeter` is implemented by using formulas for [Algorithm](https://edwilliams.org/avform.htm#Dist) & [it's NOAA online calculator](www.nhc.noaa.gov)

Created from templates made available by Stagehand under a BSD-style
[license](https://github.com/dart-lang/stagehand/blob/master/LICENSE).




CHANGELOG.md
## 1.0.0
- Initial version, with `Polygon`, `contains()`
## 1.0.1 
As suggested by pub.dartlang.org analysis
- Ran dartfmt on poly.dart & 
- Added description in pubspec.yaml
## 1.0.2
As suggested by pub.dartlang.org analysis
- Removed some description from pubspec.yaml
## 1.0.3
Added following functions & examples for - 
* [Exception handling](example/exception_handling.dart)
* [Easy Casting](example/easy_casting.dart)
* [Simple CSV](example/simple_csv.dart)
* [Conversion](example/conversion.dart)

## 1.0.4
- Optimized imports
- Added `Polygon.hasSamePoint()`
- [Issue raised for `List<List<dyanamic>>` to `List<List<num>>` casting exception](https://github.com/dart-lang/sdk/issues/36614 "Dart-lang List<List<dyanamic>> to List<List<num>> Casting Issue")

### 1.0.5
- Added `toListNum()` along with [examples](example/using_to.dart)
- Added `toListListNum()` along with [examples](example/using_to.dart)

### 1.0.6
- Updated documentation to include Travis CI [![build status](https://travis-ci.org/Sacchid/poly.svg)](https://travis-ci.org/Sacchid/poly#)

### 1.0.7
* Added following functions - 
    - `listOfList()`
    - `pointToList()` : `Point` => `List<num>`
    - `pointsToList`  : `List<Point>` => `List<List (x,y)>`
    
### 1.0.7+1
* Added following functions - 
    - `distanceInMeter` which calculates distance between two points {lat1,lon1} and {lat2,lon2} in meter
    - `isPointInsideT`
### 1.0.7+2 
* Fix to `isPointInsideT`

### 1.1.0
* migrated to null safety

example/simple_csv.dart
import 'package:poly/poly.dart';
import 'dart:io';

main() async {
  int countCSV = 0;
  Polygon location_eleven = Polygon([
    Point(18.507305, 73.806131),
    Point(18.508689, 73.806131),
    Point(18.509299, 73.805916),
    Point(18.509848, 73.805981),
    Point(18.512941, 73.806024),
    Point(18.512941, 73.803492),
    Point(18.512453, 73.802032),
    Point(18.512283, 73.800424),
    Point(18.511964, 73.798943),
    Point(18.511822, 73.798749),
    Point(18.513429, 73.79817),
    Point(18.514813, 73.797483),
    Point(18.514854, 73.796818),
    Point(18.515342, 73.796067),
    Point(18.515688, 73.795896),
    Point(18.516115, 73.795788),
    Point(18.517377, 73.79581),
    Point(18.518394, 73.794222),
    Point(18.519696, 73.795595),
    Point(18.520551, 73.796153),
    Point(18.523155, 73.799028),
    Point(18.523643, 73.800831),
    Point(18.523888, 73.802118),
    Point(18.524172, 73.80259),
    Point(18.524091, 73.80362),
    Point(18.524539, 73.804607),
    Point(18.524591, 73.808944),
    Point(18.524783, 73.809571),
    Point(18.524094, 73.811088),
    Point(18.524295, 73.81259),
    Point(18.525454, 73.814199),
    Point(18.526166, 73.814886),
    Point(18.524172, 73.817761),
    Point(18.523237, 73.818169),
    Point(18.5211, 73.822246),
    Point(18.520388, 73.823254),
    Point(18.520063, 73.824456),
    Point(18.519208, 73.827052),
    Point(18.517133, 73.826108),
    Point(18.516278, 73.824906),
    Point(18.514894, 73.824649),
    Point(18.514813, 73.824306),
    Point(18.513674, 73.824263),
    Point(18.513429, 73.824048),
    Point(18.511964, 73.824048),
    Point(18.511151, 73.824391),
    Point(18.51054, 73.824563),
    Point(18.509706, 73.825486),
    Point(18.507627, 73.826341),
    Point(18.507684, 73.827277),
    Point(18.506822, 73.827609),
    Point(18.506329, 73.826702),
    Point(18.505942, 73.825421),
    Point(18.506104, 73.82467),
    Point(18.507395, 73.823462),
    Point(18.508282, 73.822739),
    Point(18.509584, 73.821495),
    Point(18.510377, 73.820529),
    Point(18.510479, 73.820035),
    Point(18.510479, 73.819048),
    Point(18.510418, 73.818233),
    Point(18.50995, 73.815637),
    Point(18.509665, 73.814692),
    Point(18.509563, 73.813834),
    Point(18.50934, 73.811817),
    Point(18.508404, 73.810508),
    Point(18.508119, 73.809135),
    Point(18.507346, 73.807611),
    Point(18.507305, 73.806131),
  ]);
  List<Point<num>> checkPoints = [
    Point(18.5124785, 73.8206602),
    Point(18.5112654, 73.8172661),
    Point(18.5016256, 73.8094801),
    Point(18.5115046, 73.8171069),
    Point(18.5109839, 73.8154427),
    Point(18.5112721, 73.8162126),
    Point(18.5163667, 73.8075354),
    Point(18.5124489, 73.8207613),
    Point(18.5127482, 73.8189808),
    Point(18.5060939, 73.8259163),
    Point(18.5114909, 73.8164503),
    Point(18.508175, 73.822866),
    Point(18.5128381, 73.8112762),
    Point(18.5017223, 73.8094366),
    Point(18.5150109, 73.8159275),
    Point(18.5163667, 73.8075354),
    Point(18.5014018, 73.80952),
    Point(18.5112285, 73.8171876),
    Point(18.5078999, 73.8229854),
    Point(18.5135063, 73.8159908),
    Point(18.5110897, 73.8176941),
    Point(18.5107774, 73.8157407),
    Point(18.5011049, 73.8096167),
    Point(18.5128833, 73.8181274),
    Point(18.514984, 73.8159425),
    Point(18.5082853, 73.8229171),
    Point(18.5109213, 73.8172806),
    Point(18.5019752, 73.8094272),
    Point(18.5036271, 73.8557149),
    Point(18.5112384, 73.8171174),
    Point(18.5112384, 73.8171174),
    Point(18.5112384, 73.8171174),
    Point(18.5114182, 73.8170969),
    Point(18.5019349, 73.8096291),
    Point(18.5149707, 73.8154424),
    Point(18.5111738, 73.8171607),
    Point(18.5111634, 73.8171551),
    Point(18.5120293, 73.8113572),
    Point(18.5012618, 73.8095137),
    Point(18.5124395, 73.8207164),
    Point(18.503084, 73.8088517),
    Point(18.5124448, 73.8207177),
    Point(18.5112947, 73.8198235),
    Point(18.5117293, 73.8167963),
    Point(18.5142932, 73.8169554),
    Point(18.5123237, 73.8170578),
    Point(18.515029, 73.8159474),
    Point(18.5125214, 73.8086799),
    Point(18.5112435, 73.817181),
    Point(18.5124343, 73.8206081),
    Point(18.5079338, 73.8228738),
    Point(18.5117969, 73.8117694),
    Point(18.5112558, 73.817308),
    Point(18.5115868, 73.8169698),
    Point(18.5115922, 73.8169871),
    Point(18.5615843, 73.8141527),
    Point(18.5035994, 73.8557277),
    Point(18.56675, 73.8075583),
    Point(18.5542636, 73.8051996),
    Point(18.5024549, 73.8781556),
    Point(18.5541433, 73.805365),
    Point(18.5542963, 73.8053572),
    Point(18.5072059, 73.8262839),
    Point(18.5746887, 73.76487),
    Point(18.5543487, 73.8055963),
    Point(18.5542637, 73.8053565),
    Point(18.5542578, 73.8054209),
    Point(18.5611495, 73.8149408),
    Point(18.5555028, 73.8057247),
    Point(18.554192, 73.8054267),
    Point(18.507693, 73.796015),
    Point(18.5612433, 73.8145167),
    Point(18.5542637, 73.8053565),
    Point(18.5542637, 73.8053565),
    Point(18.5542637, 73.8053565),
    Point(18.5542637, 73.8053565),
    Point(18.5619854, 73.7928126),
    Point(18.5280783, 73.8303274),
    Point(18.5035991, 73.8557274),
    Point(18.5542247, 73.8053859),
    Point(18.5542579, 73.8052151),
    Point(18.5036077, 73.8557219),
    Point(18.5542637, 73.8053565),
    Point(18.5542735, 73.805281),
    Point(18.5552057, 73.8160962),
    Point(18.5036221, 73.8557353),
    Point(18.5542557, 73.8051856),
    Point(18.5036307, 73.8557221),
    Point(18.5606828, 73.8148879),
    Point(18.5542889, 73.8054483),
    Point(18.5542979, 73.8055085),
    Point(18.5542557, 73.8051856),
    Point(18.5035764, 73.8557534),
    Point(18.5542557, 73.8051856),
  ];

  String nameOfIsInsideResult = "IsInside.csv";
  String nameOfPolygon = "Polygon.csv";

  // Example of `IsInsideResultWithXY_ToCSVString`
  //* Saving `ArePointsInside Results` to "IsInside.csv"
  //  * Headers row will be added, as `includeHeader` isn't passed as `false`
  //  * And as `diffNameThanIsInside:"Example String"` is passed,
  //  * Header row will be `latitude,longitude,Example String`
  // Get `ArePointsInside` as CSV String - `csvOfIsInsideResult`
  String csvOfIsInsideResult = location_eleven.IsInsideResultWithXY_ToCSVString(
      checkPoints,
      diffNameThanIsInside: "Example String");
  // Write `String csvOfIsInsideResult` to "IsInside.csv"
  File(nameOfIsInsideResult).writeAsStringSync(csvOfIsInsideResult);
  print('${++countCSV}. ArePointsInside Results saved to IsInside.csv ');

  // Example of `toCSVString`
  //* Saving `Polygon.points` to "Polygon.csv"
  //  * Headers row will be added, as `includeHeader` isn't passed as `false`
  //  * And as `useXY` is passed as `true`
  //  * Header row will be `x,y`
  String csvOfPolygon = location_eleven.toCSVString(useXY: true);
  // Write `String csvOfIsInsideResult` to "Polygon.csv"
  File(nameOfPolygon).writeAsStringSync(csvOfPolygon);
  print('${++countCSV}. `Polygon.points` saved to Polygon.csv ');

  // Example of `csvToResult`
  //* Display 1st row : 2nd element & 3rd element of "IsInside.csv"
  //* e.g. here "longitude" and "Example String"
  //* As, previously `xY_IsInside_ToCSVString` returned String with header
  //* because optional parameter `header` was not set to false
  final IsInsideList = File(nameOfIsInsideResult).openRead();
  final l = await (csvToListOfList(IsInsideList) as Future<List<List<dynamic>>>);
  print("${++countCSV}. e.g. here ${l[0][1]} and ${l[0][2]}");

  // Example of `csvToPoly`
  //* Check if Point(18.507305, 73.806131) is inside Polygon readPolygon (Polygon from Polygon.csv)
  final readPolygon = File(nameOfPolygon).openRead();
  Polygon newPolygon = await csvToPoly(readPolygon);
  print(
      "${++countCSV}. Is Point(18.507305, 73.806131) inside Polygon (saved in Polygon.csv) : ${newPolygon.contains(18.507305, 73.806131)}");
}
